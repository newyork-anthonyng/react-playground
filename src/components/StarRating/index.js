import React, { Component } from "react";
import T from "prop-types";
import noop from "../../utility/noop";
import glamorous from "glamorous";

const Input = glamorous.input({
  WebkitAppearance: "none",
  ":focus + label": {
    border: "3px solid blue",
    borderRadius: "3px"
  }
});

class StarRating extends Component {
  constructor() {
    super();

    this.filledStar = "★";
    this.emptyStar = "☆";
  }

  handleClick = value => {
    if (this.props.disabled) return noop;

    return () => this.props.onChange(value);
  };

  handleMouseEnter = value => {
    if (this.props.disabled) return noop;

    return () => this.props.onMouseEnter(value);
  };

  handleMouseLeave = value => {
    if (this.props.disabled) return noop;

    return () => this.props.onMouseLeave(value);
  };

  renderStars = () => {
    const stars = [];
    for (let i = 0; i < this.props.size; i++) {
      const key = `${i + 1} rating`;
      stars.push(
        <li key={i}>
          <Input
            type="radio"
            id={key}
            value={key}
            name="StarRating"
            onClick={this.handleClick(i + 1)}
            onMouseEnter={this.handleMouseEnter(i + 1)}
            onMouseLeave={this.handleMouseLeave(i + i)}
            checked={i + 1 === this.props.value}
          />
          <label htmlFor={key} aria-label={key}>
            {i < this.props.value ? this.filledStar : this.emptyStar}
          </label>
        </li>
      );
    }

    return stars;
  };

  render() {
    return (
      <ul
        style={{ display: "flex", listStyle: "none", padding: 0 }}
        role="radiogroup"
        aria-label={`${this.props.value + 1} rating`}
      >
        {this.renderStars()}
      </ul>
    );
  }
}

StarRating.defaultProps = {
  value: 0,
  size: 5,
  disabled: false,
  onChange: noop,
  onMouseEnter: noop,
  onMouseLeave: noop
};

StarRating.propTypes = {
  value: T.number.isRequired,
  size: T.number.isRequired,
  disabled: T.bool.isRequired,
  onChange: T.func.isRequired,
  onMouseEnter: T.func.isRequired,
  onMouseLeave: T.func.isRequired
};

export default StarRating;
