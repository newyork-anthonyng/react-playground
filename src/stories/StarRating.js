import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';

import StarRating from '../components/StarRating';

storiesOf('StarRating', module)
  .addDecorator(withKnobs)
  .add('With knobs', () => {
    return (
      <StarRating
        value={number('Value', 0)}
        size={number('Size', 5)}
        disabled={boolean('Disabled')}
        onChange={action('clicked')}
        onMouseEnter={action('mouse entered')}
        onMouseLeave={action('mouse left')}
      />
    );
  })
  .add('Example', () => {
    class RatingContainer extends React.Component {
      constructor() {
        super();

        this.state = {
          rating: 0,
          prospectiveRating: null,
          disabled: false,
        };

        this.handleRatingChange = this.handleRatingChange.bind(this);
        this.handleMouseEnter = this.handleMouseEnter.bind(this);
        this.handleMouseLeave = this.handleMouseLeave.bind(this);
        this.renderMessage = this.renderMessage.bind(this);
      }

      handleRatingChange(rating) {
        this.setState({ rating });
      }

      handleMouseEnter(prospectiveRating) {
        this.setState({ prospectiveRating });
      }

      handleMouseLeave() {
        this.setState({ prospectiveRating: null });
      }

      renderMessage() {
        const rating = this.state.prospectiveRating || this.state.rating;
        switch (rating) {
          case 1:
            return 'Eek! Methinks not.';
          case 2:
            return `Meh. I've experienced better.`;
          case 3:
            return 'A-OK.';
          case 4:
            return `Yay! I'm a fan.`;
          case 5:
            return 'Woohoo! As good as it gets!';
          default:
            return '';
        }
      }

      render() {
        const { prospectiveRating, rating } = this.state;

        return (
          <div>
            <button onClick={() => this.setState({ disabled: !this.state.disabled }) }>
              {this.state.disabled ? 'Undisable': 'Disable'}
            </button>
            <StarRating
              value={prospectiveRating || rating}
              onChange={this.handleRatingChange}
              onMouseEnter={this.handleMouseEnter}
              onMouseLeave={this.handleMouseLeave}
              disabled={this.state.disabled}
            />
            {this.renderMessage()}
          </div>
        );
      }
    }

    return <RatingContainer />
  })
