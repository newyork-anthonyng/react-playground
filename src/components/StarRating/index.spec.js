import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import noop from '../../utility/noop';
import StarRating from './';

const defaultProps = {
  value: 0,
  size: 5,
  disabled: false,
  onChange: noop,
  onMouseEnter: noop,
  onMouseLeave: noop,
};

it('should render stars depending on its size', () => {
  const wrapper = shallow(
    <StarRating
      {...defaultProps}
      size={10}
    />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render its value correctly', () => {
  const wrapper = shallow(
    <StarRating
      {...defaultProps}
      value={3}
    />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should run callback when clicking on star', () => {
  const cb = jest.fn();
  const wrapper = shallow(
    <StarRating
      {...defaultProps}
      onChange={cb}
    />
  );
  const secondStar = wrapper.find('input[type="radio"]').at(1);

  secondStar.simulate('click');

  expect(cb).toHaveBeenCalledTimes(1);
  expect(cb.mock.calls[0][0]).toEqual(2);
});

it('should run callback when mouse enters a star', () => {
  const cb = jest.fn();
  const wrapper = shallow(
    <StarRating
      {...defaultProps}
      onMouseEnter={cb}
    />
  );
  const secondStar = wrapper.find('input[type="radio"]').at(1);

  secondStar.simulate('mouseenter');

  expect(cb).toHaveBeenCalledTimes(1);
  expect(cb.mock.calls[0][0]).toEqual(2);
});

it('should run callback when mouse leaves a star', () => {
  const cb = jest.fn();
  const wrapper = shallow(
    <StarRating
      {...defaultProps}
      onMouseLeave={cb}
    />
  );
  const secondStar = wrapper.find('input[type="radio"]').at(1);

  secondStar.simulate('mouseleave');

  expect(cb).toHaveBeenCalledTimes(1);
  expect(cb.mock.calls[0][0]).toEqual(2);
});
