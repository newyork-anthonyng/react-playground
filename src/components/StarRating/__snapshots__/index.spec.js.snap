// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render its value correctly 1`] = `
<ul
  aria-label="4 rating"
  role="radiogroup"
  style={
    Object {
      "display": "flex",
      "listStyle": "none",
      "padding": 0,
    }
  }
>
  <li>
    <glamorous(input)
      checked={false}
      id="1 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="1 rating"
    />
    <label
      aria-label="1 rating"
      htmlFor="1 rating"
    >
      ★
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="2 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="2 rating"
    />
    <label
      aria-label="2 rating"
      htmlFor="2 rating"
    >
      ★
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={true}
      id="3 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="3 rating"
    />
    <label
      aria-label="3 rating"
      htmlFor="3 rating"
    >
      ★
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="4 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="4 rating"
    />
    <label
      aria-label="4 rating"
      htmlFor="4 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="5 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="5 rating"
    />
    <label
      aria-label="5 rating"
      htmlFor="5 rating"
    >
      ☆
    </label>
  </li>
</ul>
`;

exports[`should render stars depending on its size 1`] = `
<ul
  aria-label="1 rating"
  role="radiogroup"
  style={
    Object {
      "display": "flex",
      "listStyle": "none",
      "padding": 0,
    }
  }
>
  <li>
    <glamorous(input)
      checked={false}
      id="1 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="1 rating"
    />
    <label
      aria-label="1 rating"
      htmlFor="1 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="2 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="2 rating"
    />
    <label
      aria-label="2 rating"
      htmlFor="2 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="3 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="3 rating"
    />
    <label
      aria-label="3 rating"
      htmlFor="3 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="4 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="4 rating"
    />
    <label
      aria-label="4 rating"
      htmlFor="4 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="5 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="5 rating"
    />
    <label
      aria-label="5 rating"
      htmlFor="5 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="6 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="6 rating"
    />
    <label
      aria-label="6 rating"
      htmlFor="6 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="7 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="7 rating"
    />
    <label
      aria-label="7 rating"
      htmlFor="7 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="8 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="8 rating"
    />
    <label
      aria-label="8 rating"
      htmlFor="8 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="9 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="9 rating"
    />
    <label
      aria-label="9 rating"
      htmlFor="9 rating"
    >
      ☆
    </label>
  </li>
  <li>
    <glamorous(input)
      checked={false}
      id="10 rating"
      name="StarRating"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="radio"
      value="10 rating"
    />
    <label
      aria-label="10 rating"
      htmlFor="10 rating"
    >
      ☆
    </label>
  </li>
</ul>
`;
